RFC: device: convert device struct instances to const

pabigot opened the pull request:
This PR has an alternative prototype of the part of #24873 that allows struct device instances to be put into flash.  The intent of this draft PR is to examine the impact of declaring device structures with a `const` qualifier.

A simple representation of generic mutable driver state ("context") was selected and named to enable evaluation of the impact of making device structures immutable.  The final approach to associating context with drivers will be determined in other contexts including #24873.

Initial results reveal the significance of the change:
* Every device driver init function and sys_init function must be updated to take `const struct device*` where `struct device*` is currently passed.  This can be done relatively easily using a Coccinelle script for code, and a sed script for documentation and macros where Coccinelle can't detect the pattern.
* The passing of `const device *` pointers to functions with `void *` parameters is a much bigger issue; at this time the situations can only be detected by build failures.  Many are currently being masked in this draft by a commit that removes the `const` qualifier from the device pointer produced by `DEVICE_GET()`.  Device pointers are sometimes passed:
  - as parameters to thread handlers (any of the three `void *` arguments)
  - to ISR handlers (generally unary functions, but not always)
  - as context to `log_output_ctx_set()`
* Some changes are also required in external modules, e.g. the `SYS_INIT()` used in the Segger RTT capability.

Commit:
drivers/gpio: Use DEVICE_AND_API_INIT for mmio32 driver

Let's set the api at built time, or this will create a bug once device
instance pointers become constant.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

pabigot committed:
device: add a common RAM context structure for devices

At this time the only thing preventing moving `struct device` objects
into flash is the use of the API pointer as a flag indicating
initialization success.  Add a structure to hold dynamic state common
to all devices.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
device: switch to new record of successful initialization

Now that struct device instances are immutable store the flag that
indicates whether initialization was successful in the device context.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
device: use helper macro in device define macros

Avoid duplication of the encoding of the public device identifier by
using the constructing macro in definitions.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
DNM WIP device: add illegal cast to dev pointer passed as userdata

Various generic data pointers are typed as pointer-to-void but some
uses pass device pointers.  Add const-removing casts to avoid the
build errors.

Some of these were identified from shippable logs, e.g.:

grep discarded-qualif /tmp/zephyrproject-rtos-zephyr-72591.1.log \
  | sed -e 's@^.*/@@' \
  | sort \
  | uniq

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
device: define struct device instances as const

These should now have no mutable content.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot commented:
Existing code that passes a device pointer as an argument to a function that receives a `void *` parameter is going to be a huge problem.  The `const` qualifier must be cast away in the ISR infrastructure, as well as many thread entry functions and other callback user-data parameters.  At this time the necessary conversions don't appear to be automatable.

tbursztyka commented on include/device.h:
```diff
@@ -254,6 +271,7 @@ struct device {
 	const void *config_info;
 	const void *driver_api;
 	void * const driver_data;
+	struct device_context * const driver_context;
```
I'd just call it context (imo all these attributes should be renamed: config_info -> config, driver_api -> api, driver_data -> data), but that's a side note.

What's the benefit exposing it that way? (adding a pointer in the struct, instead of a fully separate array as in #24873).  It's a trade-off between ROM occupation and runtime computation, anything else than that?

In the very end, we should merge driver_data and driver_context, and provide the means for the drivers to access the data seamlessly (via a macro or else). I just did not go myself in that path in my RFC because it generates a lot of changes in drivers

pabigot replied:
Exactly as you say: the right way is to put the common context into a prefix of the driver-specific data.  That can't be done if they're separate objects.

This is a prototype using a different object relationship than yours; we don't know yet what the best path will be.  Better to try several and compare their pros and cons.

tbursztyka replied:
>  That can't be done if they're separate objects.

Of course it can be. Current RFCs (mine and yours) are only trying to solve the device instance as constant. 

Once done _then_ we can move on with modifying driver_data to provide also a driver context. 

But, as I said in my other comment, if the goal is _only_ to solve the device instances to be constants, there is 0 point to introduce right now a struct device_context since an atomic bits array would just do the work.
In my RFC I did introduce such context because of all the other features (sync/lock and status reporting), not because of const-ifying devices. 

So maybe we should really proceed step by steps: 
- let's fix the device constant by having such atomic bits (and let device_get_binding() check these when relevant)
- let's add a struct device_context for everything else that would come afterwards (sync/lock & status reporting).

As you already mentioned in relevant PRs, you are not convinced about adding such sync/lock and status report features... so adding a device context right now would be a bad idea then.

pabigot replied:
You're right, I was combining multiple features.  I'll switch it to a bool `init_ok` since that's the only state we need right now.

tbursztyka commented on subsys/shell/modules/device_service.c:
```diff
@@ -39,7 +39,7 @@ static bool device_get_config_level(const struct shell *shell, int level)
 
 	for (dev = levels[level]; dev < levels[level+1]; dev++) {
-		if (dev->driver_api != NULL) {
+		if (dev->driver_context->init_ok) {
```
hm, if we stick to the goal of this RFC: const-ifying struct device instances, then such driver context is overkill.

Better having an array of atomic bits then, like PM's busy bit array. Less changes to reach the goal.

tbursztyka commented:
perhaps cherry-pick the obvious patches I have already done in #24873 about putting struct device to const instead of redoing them.
you'll need that one as well https://github.com/zephyrproject-rtos/zephyr/pull/24873/commits/37171961616717471ffc455a92fe9b3000153533

pabigot commented:
> perhaps cherry-pick the obvious patches I have already done in #24873 about putting struct device to const instead of redoing them.

I looked at them and this told me it wasn't right:
```
                if (dev != NULL) {
-                       z_object_init(dev);
+                       z_object_init((void *)dev);
                }
```
We can't go removing const from pointers.  It's outlawed by MISRA 11.8:
> A cast shall not remove any const or volatile qualification from the type pointed to by a pointer

tbursztyka commented:
> We can't go removing const from pointers. It's outlawed by MISRA 11.8:

Did not know that, I should probably download someday the misra papers.

Anyway my point was that instead of chasing the same rabbit in parallel, let's use each-others works properly. At least, I am not going to waste my time on a task if I know somebody else is doing the same work. Unless, you want to take over this task (though it makes already my work a bit of a waste then)

pabigot commented:
I'm not interested in taking over your wider work on the device structure, but I am interested in pursuing issues related to making `struct device` const.  I don't feel that this work conflicts with yours, but I do feel that trying to isolate those issues in the context of a set of draft patches that touch many files is too difficult.  So I'm trying to do the minimum necessary to isolate those specific issues.

I expect to have my results in the next day or two after a few more passes through shippable to identify those issues, then we can assess how to proceed.  This effort has already identified existing MISRA violations in current master that should probably be addressed before 2.3 release (#25247).

pabigot committed:
MANUAL device: ensure all device pointers are to const objects

Rebases should pause immediately before this commit and run the steps
described below, which should be committed with `-C SHA` where SHA is
is the commit SHA of this commit.  This may reduce the amount of
conflict arising from upstream changes.

1: coccinelle: ensure all code references point to const struct device

/* Force struct device to be const in all pointer declarations
 * Invoke:

spatch --sp-file devptr.cocci --dir . --include-headers --very-quiet \
    --macro-file scripts/coccinelle/macros.h

 */

@
disable optional_qualifier
@
@@
+const
 struct device *

@
disable optional_qualifier
@
@@
+const
 struct device * const

2: sed: ensure all non-code/macro references point to const struct device

git grep -l '(struct device' \
  | xargs sed -i -r \
    -e 's@\(struct device@(const struct device@g'

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
DNM device: cast const away from DEVICE_GET() value

This macro is generally used in contexts where something is expecting
a pointer to (non-const) void.  Reduce the complaints by lying about
the type.

This needs to be fixed properly by changing all API where an
application of this macro is passed as a parameter; i.e. specifically
API related to ISRs.  Here we're just testing build and behavior.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
userspace: update k_object API to support immutable objects

The k_object API associates mutable state structures with known kernel
objects to support userspace.  The kernel objects themselves are not
modified by the API, and in some cases (e.g. device structures) may be
const-qualified.  Update the API so that pointers to these const
kernel objects can be passed without casting away the const qualifier.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
DNM west: update to repo versions that support new API

External modules may include code with prototypes that must be updated
to accept const device objects.  Update to the version of those
modules that is ready for the new expectations.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
drivers: use macro to define device structures

Replace individual device instance definitions with the macro that
expands to the equivalent change.

    git grep -l 'struct device DEVICE_NAME_GET' \
     | xargs sed -i -r \
       -e 's@static struct device DEVICE_NAME_GET\(([^)]*)\);@DEVICE_DECLARE(\1);@'

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

galak commented:
Dev-Review (May 14):
* We should take a step back and decide if making device struct const is a direction we want to go, and thus preclude runtime creation of devices.  (@galak the 2 examples I can come up with are PCI/e devices and USB, scanning a bus at runtime)

pabigot committed:
drivers: i2c: dw: remove set of driver_api pointer

This is done by the init infrastructure and should not be implemented
in the driver itself.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot commented:
Latest run passes all tests, so we'll leave it here.  Some of the commits should be submitted for master regardless of whether we continue with making definitions const.
